collections/collections.go

```diff
@@ -0,0 +1,98 @@
+package collections
+
+import (
+	"bytes"
+	"fmt"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// ErrNotFound defines an error returned when
+// a key was not found.
+type ErrNotFound struct {
```

Comment:
we could simply use an fmt.Errorf vs using a concrete struct here, but I can see value in allowing to parse the error contents programmatically

Reply:
@aaronc wdyt of this? should we extend it to contain also the module the not found error came from? since we have the store key. I think it might bring value to have a full error saying "object V of module M was not found: key K"

Reply:
The most important thing IMO is being able to check if it's a not found error, i.e. `collections.IsNotFound(err)`

Reply:
Oki, will add. WDYT about implementing gRPC status on it too?

Reply:
It's unlikely that consumers of this package will be able to have meaningfully different behaviors based on details within a NotFound error. +1 to just defining `var ErrNotFound = errors.New("not found")` and returning it as a value rather than a type.

Reply:
Taking into consideration that the error , in 99% of cases, is going to probably end up to be inspected after going through the wire, I am leaning on just having a simple `ErrNotFound`

Reply:
Go errors don't go over the wire directly. But, yes, we agree, I think: you want `var ErrNotFound = errors.New(...)` and not `type ErrNotFound struct { ... }` (or actually should be `NotFoundError`).

Reply:
> Go errors don't go over the wire directly.

Yes this is why I was saying it's better to have it as a value rather than a type, since wire does not have context on go types, and concretely speaking this value will be most likely returned during gRPC queries so a reason more not to define a type because all the structured context will be lost.

Reply:
Addressed.

@aaronc, since ErrNotFound is now a value I think using the std `errors.Is` should be enough.